Certificate provider (#130)
AVSystem/developer-zone: docs/LwM2M_Client/Nordic/Runtime_certificate.md

    west config manifest.file west-nrf.yml
    west update
    ```

0. Build a project with a runtime certificate and private key.

    This feature works with nRF9160DK starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:

    ```
    west build -b nrf9160dk_nrf9160_ns@0.14.0 -p -- -DCONF_FILE=prj_extflash.conf -DOVERLAY_CONFIG="overlay_nrf_mbedtls.conf"
    ```

0. Flash the board with `west flash` command.

## Generate certificate
The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate them open terminal in the **Anjay-zephyr-client/demo** directory and use the following commands.

```
openssl ecparam -name secp256r1 -out ecparam.der
openssl req -new -x509 -nodes -newkey ec:ecparam.der -keyout demo-cert.key -out cert.pem -days 3650
openssl ec -in demo-cert.key -outform pem -out key.pem
```

![Fragment of creating certificates](images/create_cert.png)

!!! Important

    To use the certificate and private key with Coiote IoT DM you must specify a common name that is the same as the client endpoint name.

> You will see created `cert.pem` and `key.pem` files in the `Anjay-zephyr-client/demo` directory.

Reviewer: Something is grammatically not right in this sentence (I think). 
```suggestion
You will see two files, `cert.pem` and `key.pem` created in the `Anjay-zephyr-client/demo` directory.
```

## Configure the Client

> 0. Provide the generated certificate and private key through the shell in the serial communication program.

Reviewer: I'm not sure how mistake-proof should this guide be. Should we say how to connect to the shell using the serial communication program? I fell that the nRF9160 and Thingy91 descriptions how to set psk is made with more details.

    !!! Note

        Before setting configurations in Anjay you need to stop running Anjay by `anjay stop` command.

    Run these commands and paste generated in the previous step files.

    ```
    anjay config set public_cert
    anjay config set private_key
    ```

    ![Generate certificate in Anjay](images/anjay_cert.png)

0. Save changes, start Anjay using `anjay start` command and go to the next step to add a device to Coiote.

## Add device to Coiote IoT DM

0. Upon logging in to Coiote IoT DM for the first time, you will see the **Add your LwM2M device** panel.

    !!! note
        If you had previously added a device, in **Device inventory**, click the **Add device** button.

0. Select the **Connect your LwM2M device directly via the Management server** tile.
   ![Add via Mgmt](images/ex1.3.png "Add via Mgmt")
0. In the **Device credentials** step:
     - In the **Endpoint name** enter your LwM2M device endpoint name, e.g. `anjay-zephyr-demo`.
     - In the **Security mode** section, select **Certificate** mode.
        ![Device credentials step](images/add_mgmt_cert.png "Device credentials step")
     - Click **Upload a new certificate** and **Browse**.

>      - In the pop-up, go to the directory where your certificate has been generated, select the `demo-cert.crt` file and click **Open**.

Reviewer: ```suggestion
     - In the pop-up, go to the directory where your certificate has been generated, select the `cert.pem` file and click **Open**.
```
Currently the file created is called cert.pem not demo-cert.crt

# Runtime Certificate and Private Key Configuration

## Introduction
Anjay library supports secure connection between device and server. Encryption can be achieved with the use of Pre-Shared Key as it was described in [nRF9160](nRF9160.md) and [Thingy:91](Thingy91.md) pages or with certificate-based cryptography.

> In **Certificate Mode** an asymmetrical public-key cryptographic algorithm is used to authenticate the connection endpoints and initialize payload encryption. Appropriate certificates need to be generated for both the LwM2M Client and the LwM2M Server. Public certificates of both parties are mutually available, and each party also has access to its corresponding private key.

Reviewer 2: It's either `Asymmetric Algorithm` or `Public Key Cryptography`.

> Read more about secure communication on [Anjay's documentation](https://avsystem.github.io/Anjay-doc/BasicClient/BC-Security.html) page.

Reviewer 2: ```suggestion
You can read more about secure communication on [Anjay's documentation](https://avsystem.github.io/Anjay-doc/BasicClient/BC-Security.html) page.
```

## Prerequisites

> * The nRF9160 board with a USB cable.

Reviewer 2: nRF9160 is a chip, nRF9160DK is a board we use. I think the whole DevZone should be updated.

Reviewer 2: Also, there should be an information that other boards can be used and nrf9160dk is used here as an example.

- Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
* Installed **minicom** (for Linux), **RealTerm**, **PuTTy** (for Windows), or another serial communication program.
* An active [Coiote IoT DM](https://eu.iot.avsystem.cloud/) user account.

## Build and flash the device
0. Connect the nRF9160 board to a USB port of your machine.
0. Go to your local **Anjay-zephyr-client** directory
0. Set West manifest path to `Anjay-zephyr-client/demo`, manifest file to `west-nrf.yml`, and do `west update`:

    ```
    west config manifest.path Anjay-zephyr-client/demo
    west config manifest.file west-nrf.yml
    west update
    ```

0. Go to the **Anjay-zephyr-client/demo** folder and build a project with a runtime certificate and private key.

    ```
    west build -b <BOARD> -p -- -DCONFIG_ANJAY_ZEPHYR_RUNTIME_CERT_CONFIG=y
    ```

    where <BOARD\> should be replaced by the selected board from `boards/` directory.

    !!! Note

        The runtime certificate and private key do not work with Thingy:91. Other boards should work with this command.

>     This feature works with nrf9160dk starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:

Reviewer 2: ```suggestion
    This feature works with nRF9160DK starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:
```

    ```
    west build -b nrf9160dk_nrf9160_ns@0.14.0 -p -- -DCONF_FILE=prj_extflash.conf -DOVERLAY_CONFIG="overlay_nrf_mbedtls.conf"
    ```

0. Find the merged.hex file under the `build/zephyr` directory in the project folder.

> 0. Use the nRF Connect Programmer with the downloaded `.hex` file and execute steps from the [Updating the application firmware](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_nrf9160_gs.html#updating-the-application-firmware) section.

Reviewer 2: Why?????? `west flash` does the work here...

## Generate certificate

> The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate them open terminal and use the following commands.

Reviewer 2: ```suggestion
The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate them open terminal, go to the `demo/` direcotry and use the following commands.
```

Author: but we are in demo folder

Reviewer 2: My bad, I would specify it explicitly that we should do it in the `demo/` directory anyway. Just for clarification for the user.

!!! Important

>     To use the certificate and private key with Coiote IoT DM you must specify a common name that is the same as the client endpoint name.

Reviewer 2: Where do I specify that? I've never done that and seriously I have no idea.

Author: by this commands. You create the certificate by typing informations in the terminal.

Reviewer 2: But where is the place where I provide the `common name that is the same as the client endpoint name`., Your endpoint name here is the `anjay-zephyr-demo` but I don't see it in the example commands etc. I'm seriously confused.

```

> openssl ecparam -name secp256r1 -out ecparam.der

Reviewer 2: shouldn't `openssl` be in the prerequisites?

openssl req -new -x509 -nodes -newkey ec:ecparam.der -keyout demo-cert.key -out demo-cert.crt -days 3650
openssl x509 -in demo-cert.crt -outform pem -out cert.pem
openssl ec -in demo-cert.key -outform pem -out key.pem
```

![Fragement of creating certificates](images/create_cert.png)

You will see created `demo-cert.pem` and `demo-cert.key.pem` files in the `Anjay-zephyr-client/demo` directory.

## Configuring the Client
0. Provide the generated certificate and private key through the shell in the serial communication program.

    !!! Note

        Before setting configurations in Anjay you need to stop running Anjay by `anjay stop` command.

    ```
    anjay config set public_cert
    anjay config set private_key
    ```

    ![Generate certificate in Anjay](images/anjay_cert.png)

> 0. Save created changes, start Anjay and go to the next step to add a device to Coiote.

Reviewer 2: ```suggestion
0. Save changes, start Anjay using `anjay start` command and go to the next step to add a device to Coiote.
```

> ## Add device to Coiote IoT DM and connect

Reviewer 2: ```suggestion
## Add device to Coiote IoT DM
```

0. Upon logging in to Coiote IoT DM for the first time, you will see the **Add your LwM2M device** panel.

    !!! note
        If you had previously added a device, in **Device inventory**, click the **Add device** button.

0. Select the **Connect your LwM2M device directly via the Management server** tile.
   ![Add via Mgmt](images/ex1.3.png "Add via Mgmt")
0. In the **Device credentials** step:

>      - In the **Endpoint name** enter your LwM2M device endpoint name, e.g. `anjay-zephyr-demo`.

Reviewer 2: add a note that it should be the same as in the zephyr shell

* An active [Coiote IoT DM](https://eu.iot.avsystem.cloud/) user account.

!!! Note

    The runtime certificate and private key configuration do not work with other boards.

## Build and flash the device
0. Connect the nRF9160 DK board to a USB port of your machine.
0. Enter the command line interface on your machine and change the directory to the one where you set up the Zephyr project (usually, it's `~/zephyrproject`).
0. Go to your local **Anjay-zephyr-client/demo** directory
0. Set manifest file to `west-nrf.yml`, and do `west update`:

    ```
    west config manifest.path Anjay-zephyr-client/demo
    west config manifest.file west-nrf.yml
    west update
    ```

0. Build a project with a runtime certificate and private key.

    This feature works with nRF9160 DK starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:

    ```
    west build -b nrf9160dk_nrf9160_ns@0.14.0 -p -- -DCONF_FILE=prj_extflash.conf -DOVERLAY_CONFIG="overlay_nrf_mbedtls.conf"
    ```

0. Flash the board with `west flash` command.

## Generate certificate

> The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate To generate a self sign certificate and key pair open a terminal in the **Anjay-zephyr-client/demo** directory and use the following commands.

Reviewer 3: ```suggestion
The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate To generate a self-signed certificate and key pair open a terminal in the **Anjay-zephyr-client/demo** directory and use the following commands.
```

```
openssl ecparam -name secp256r1 -out ecparam.der
openssl req -new -x509 -nodes -newkey ec:ecparam.der -keyout demo-cert.key -out cert.pem -days 3650
openssl ec -in demo-cert.key -outform pem -out key.pem
```

![Fragment of creating certificates](images/create_cert.png)

!!! Important

    To use the certificate and private key with Coiote IoT DM you must specify a common name that is the same as the client endpoint name.

You will see two files `cert.pem` and `key.pem` created in the `Anjay-zephyr-client/demo` directory.

## Configure the Client
0. Provide the generated certificate and private key through the shell in the serial communication program.

    !!! Note

        Before setting configurations in Anjay you need to stop running Anjay by `anjay stop` command.

>     Run these commands and paste generated in the previous step files.

Reviewer: nit. 
```suggestion
    Run these commands and paste the content of files generated in the previous step.
```

# Runtime Certificate and Private Key Configuration

## Introduction
Anjay library supports secure connection between device and server. Encryption can be achieved with the use of Pre-Shared Key as it was described in [nRF9160](nRF9160.md) and [Thingy:91](Thingy91.md) pages or with certificate-based cryptography.

In **Certificate Mode** an asymmetrical algorithm is used to authenticate the connection endpoints and initialize payload encryption. Appropriate certificates need to be generated for both the LwM2M Client and the LwM2M Server. Public certificates of both parties are mutually available, and each party also has access to its corresponding private key.

You can read more about secure communication on [Anjay's documentation](https://avsystem.github.io/Anjay-doc/BasicClient/BC-Security.html) page.

## Prerequisites
* The nRF9160 DK board with a USB cable.
* Cloned [Anjay-zephyr-client](https://github.com/AVSystem/Anjay-zephyr-client#getting-started) repository

!!! Note
    Go through the README's `Getting Started` part.

> * Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)

Reviewer 4: Is that explicitly needed? I think this is installed while doing 
```
   west config manifest.file west-nrf.yml
    west update
 ```

Author: Let's leave this as it was. It's better to have this than fighting with possible problems later

## Prerequisites
* The nRF9160DK board with a USB cable.
* Cloned [Anjay-zephyr-client](https://github.com/AVSystem/Anjay-zephyr-client#getting-started) repository
* Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
* Installed [OpenSSL](https://www.openssl.org/source/)
* Installed **minicom** (for Linux), **RealTerm**, **PuTTy** (for Windows), or another serial communication program.
* An active [Coiote IoT DM](https://eu.iot.avsystem.cloud/) user account.

!!! Note
    In this tutorial we will use the nRF9160DK board as an example.

## Build and flash the device
0. Connect the nRF9160DK board to a USB port of your machine.
0. Go to your local **Anjay-zephyr-client/demo** directory
0. Set manifest file to `west-nrf.yml`, and do `west update`:

    ```
    west config manifest.file west-nrf.yml
    west update
    ```

0. Build a project with a runtime certificate and private key.

    ```
    west build -b <BOARD> -p -- -DCONFIG_ANJAY_ZEPHYR_RUNTIME_CERT_CONFIG=y
    ```

>     where <BOARD\> should be replaced by the selected board from `boards/` directory.

Reviewer: We are assuming that we use nRF9160DK so this <BOARD> parameter can be replaced by the real value.

    !!! Note

        The runtime certificate and private key configuration do not work with Thingy:91. Other boards should work with this command.

    This feature works with nRF9160DK starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:

    ```

>     west build -b nrf9160dk_nrf9160_ns@0.14.0 -p -- -DCONF_FILE=prj_extflash.conf -DOVERLAY_CONFIG="overlay_nrf_mbedtls.conf"

Reviewer: It's a little confusing should we run this build command or the previous one where we set DCONFIG_ANJAY_ZEPHYR_RUNTIME_CERT_CONFIG=y .Both of those two commands are described in the same point.

Reviewer 4: I agree, I don't understand that point

    ```

0. Flash the board with `west flash` command.

## Generate certificate

> The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate them open terminal in the **Anjay-zephyr-client/demo** directory and use the following commands.

Reviewer: ```suggestion
The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate a self sign certificate and key pair open a terminal in the **Anjay-zephyr-client/demo** directory and use the following commands.
```

Reviewer: Proposal rejected :cry: ?

Reviewer: Also I just noticed that in `overlay_nrf_mbedtls.conf` we configure support for both EC P256 and EC P384 curves. I know that this description is based on the README file in Anjay-Zephyr-client but did anyone verified that we support  only P256? Maybe we support P384 also and could say that the user can choose one of those curves?

Author: sorry I thought I added this

Reviewer 3: `self-signed` btw

Reviewer: About the P384 support. Discussed offline that this can stay as is.

```
openssl ecparam -name secp256r1 -out ecparam.der
openssl req -new -x509 -nodes -newkey ec:ecparam.der -keyout demo-cert.key -out demo-cert.crt -days 3650

> openssl x509 -in demo-cert.crt -outform pem -out cert.pem

Reviewer: This command literally only copies the file with the new name! The demo-cert.crt and cert.pem are the same! The reason for that is demo-cert.crt is already in pem format

openssl ec -in demo-cert.key -outform pem -out key.pem
```

![Fragment of creating certificates](images/create_cert.png)

!!! Important

    To use the certificate and private key with Coiote IoT DM you must specify a common name that is the same as the client endpoint name.

You will see created `cert.pem` and `key.pem` files in the `Anjay-zephyr-client/demo` directory.

## Configure the Client
0. Provide the generated certificate and private key through the shell in the serial communication program.

    !!! Note

        Before setting configurations in Anjay you need to stop running Anjay by `anjay stop` command.

    ```
    anjay config set public_cert

>     anjay config set private_key

Reviewer: Probably it would be nice to say call this command and paste the content of the .pem file or something.

In **Certificate Mode** an asymmetrical algorithm is used to authenticate the connection endpoints and initialize payload encryption. Appropriate certificates need to be generated for both the LwM2M Client and the LwM2M Server. Public certificates of both parties are mutually available, and each party also has access to its corresponding private key.

You can read more about secure communication on [Anjay's documentation](https://avsystem.github.io/Anjay-doc/BasicClient/BC-Security.html) page.

## Prerequisites
* The nRF9160DK board with a USB cable.
* Cloned [Anjay-zephyr-client](https://github.com/AVSystem/Anjay-zephyr-client#getting-started) repository
* Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
* Installed [OpenSSL](https://www.openssl.org/source/)
* Installed **minicom** (for Linux), **RealTerm**, **PuTTy** (for Windows), or another serial communication program.
* An active [Coiote IoT DM](https://eu.iot.avsystem.cloud/) user account.

!!! Note

>     In this tutorial we will use the nRF9160DK board as an example.

Reviewer 3: We should verify if any board other than nRF9160 DK with external flash can fit Anjay demo with this option turned on...

## Build and flash the device
0. Connect the nRF9160DK board to a USB port of your machine.
0. Go to your local **Anjay-zephyr-client** directory
0. Set West manifest path to `Anjay-zephyr-client/demo`, manifest file to `west-nrf.yml`, and do `west update`:

    ```

>     west config manifest.path Anjay-zephyr-client/demo

Reviewer 3: If we just applied
`Go to your local **Anjay-zephyr-client** directory`
now the correct path would be just `demo`. Since in the next step we navigate to `demo` directory anyway, maybe go there in the first place and skip `west config manifest.path`?

Reviewer 3: bump

    west config manifest.file west-nrf.yml
    west update
    ```

0. Build a project with a runtime certificate and private key.

    ```
    west build -b <BOARD> -p -- -DCONFIG_ANJAY_ZEPHYR_RUNTIME_CERT_CONFIG=y
    ```

    where <BOARD\> should be replaced by the selected board from `boards/` directory.

    !!! Note

        The runtime certificate and private key configuration do not work with Thingy:91. Other boards should work with this command.

>     This feature works with nRF9160DK starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:

Reviewer 4: What about other boards?

# Runtime certificate and private key configuration

> Build a project with a runtime certificate and private key.

Reviewer 4: ```suggestion
## Introduction
Anjay library supports secure connection between device and server. Encryption can be achieved with use of Pre-Shared Key as it was described in
[nRF9160](nRF9160.md) and [Thingy:91](Thingy91.md) pages or with certificate based cryptography. In **Certificate Mode** an asymmetrical public-key cryptographic algorithm is used to authenticate the connection endpoints and initialize payload encryption. Appropriate certificates need to be generated for both the LwM2M Client and the LwM2M Server. Public certificates of both parties are mutually available, and each party also has access to its own corresponding private key.

Read more about secure communication on [Anjay-doc](https://avsystem.github.io/Anjay-doc/BasicClient/BC-Security.html) page.

Here’s a tutorial to show you how to compile Anjay configured to use certificates, generate them, upload to the device and connect to [Coiote IoT DM](https://eu.iot.avsystem.cloud/).

```

## Prerequisites
* The nRF9160 board with a USB cable.
- Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
* Installed **minicom** (for Linux), **RealTerm**, **PuTTy** (for Windows), or another serial communication program.
* An active [Coiote IoT DM](https://eu.iot.avsystem.cloud/) user account.

## Build and flash the device
0. Connect the nRF9160 board to a USB port of your machine.

> 0. Go to your local **Anjay-zephyr-client** directory

Reviewer 4: Since we do not explain how to clone repository we should add another requirements to `Prerequisites` to prepare cloned repository.

Author: The link moves you to the README so we don't need to explicitly point this out

Reviewer 4: Someone can think that only following prerequisites on devzone is enough(It was me :grinning: ) so in my opinion pointing out to follow all `README` steps and remove from prerequisities those which are doubled is a good option

0. Set West manifest path to `Anjay-zephyr-client/demo`, manifest file to `west-nrf.yml`, and do `west update`:

    ```
    west config manifest.path Anjay-zephyr-client/demo
    west config manifest.file west-nrf.yml
    west update
    ```

0. Go to the **Anjay-zephyr-client/demo** folder and build a project with a runtime certificate and private key.

    ```
    west build -b <BOARD> -p -- -DCONFIG_ANJAY_ZEPHYR_RUNTIME_CERT_CONFIG=y
    ```

>     where <BOARD\> should be replaced by the selected board from boards/ directory.

Reviewer 4: ```suggestion
    where <BOARD\> should be replaced by the selected board from `boards/` directory.
```

    !!! Note

        The runtime certificate and private key do not work with Thingy:91. Other boards should work with this command.

    This feature works with nrf9160dk starting from revision v0.14.0. For this board use configuration that utilizes an external flash chip and software-based cryptography:

    ```
    west build -b nrf9160dk_nrf9160_ns@0.14.0 -p -- -DCONF_FILE=prj_extflash.conf -DOVERLAY_CONFIG="overlay_nrf_mbedtls.conf"
    ```

> 0. Find the merged.hex file under the **build/zephyr** directory in the project folder.

Reviewer 4: ```suggestion
0. Find the merged.hex file under the `build/zephyr` directory in the project folder.
```

0. Use the nRF Connect Programmer with the downloaded `.hex` file and execute steps from the [Updating the application firmware](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_nrf9160_gs.html#updating-the-application-firmware) section.

## Generate certificate

> The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in PEM format. To generate them open terminal and use the following commands.

Reviewer 4: ```suggestion
The certificate and private key based on the SECP256R1 curve can be provided through the shell interface in `.pem` format. To generate them open terminal and use the following commands.
```
